Refactor chain state persistency

ffakenz opened the pull request:
<!-- Describe your change here -->

🪚 Remove previous attribute from ChainStateAt.

🪚 Since ChainStateType is not recursive anymore, the HeadState now keeps a non-empty list of ChainStateType instead.

🪚 This means, now the HeadLogic has to collect the new ChainStateType during 'aggregate'.

🪚 Finally, the LocalChainState now has to maintain a non-empty list instead in order to implement 'rollback'.

🪚 As a consequence, now the persistent state is more lightweight and easier to read and work with.

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ffakenz committed:
Remove previous from ChainStateAt

Since ChainStateType is not recursive anymore, the HeadState now keeps a
non-empty list of ChainStateType instead.

This means, now the HeadLogic has to collect the new ChainStateType during
'aggregate'.

Finally, the LocalChainState now has to maintain a non-empty list instead
in order to implement 'rollback'.

ffakenz committed:
Fix golden specs

ffakenz committed:
Update CHANGELOG

ffakenz committed:
Fix handler spec for LocalChainState

Introduce 'history' handle to get all chain changed events.
Note: in the future we might want to change the interface to be a stream.

This new handle was needed in order to be able to resume the LocalChainState.

ffakenz committed:
Fix Rollback to contain the rolledback history of chain states

We were aggregating wrong the Rollback observation by appending the rolledBack
point to the previous history, but we needed to override it instead.

v0d1ch commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -67,40 +68,54 @@ import System.IO.Error (userError)
-      | recordPoint <= rollbackChainPoint -> cs
-    ChainStateAt{previous = Just prev} -> go rollbackChainPoint prev
-    cs -> cs
+  go rollbackChainPoint history =
```
We usually name functions `go` when they are recursive. Now this one is not anymore so maybe we could give it more descriptive name?

v0d1ch reviewed (approved):
:tada:

ffakenz committed:
Make use of 'cons' symbol to add elemnts to NonEmpty list

ffakenz committed:
Introduce newtype ChainStateHistory

Add helper functions to avoid scattering the code all over the place.
Also made some renaming to keep things cohesive

ffakenz committed:
Minor formatting changes

ffakenz committed:
Fix golden specs

pgrange reviewed (approved):
Before, the chainState was really abstract to the HeadLogic but that came with the overhead of having to store the full chainState history with every chain originated event. I'm not sure how heavy this overhead was but I can see how this PR removes it and that's cool.

So now, we do not store the chainState full history every time but it comes with the HeadlLogic having to know about the fact that there is a notion of chainState history and the headLogic know has to  deal with it and, for instance, call this `pushNewState` function.

That's a shame the HeadLogic has to have this knowledge but I understand that there is a trade-off here and I'm fine with this way of dealing with it 👍

ffakenz committed:
Add function to recover ChainStateHistory from loaded events

This removes the need to keep the entire history as part of the HeadState,
keeping the persistence size at minimum.

ffakenz committed:
Fix golden specs

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -220,7 +248,7 @@ data ChainEvent tx
       }
   | Rollback
-      { rolledBackChainState :: ChainStateType tx
+      { rolledBackChainState :: ChainStateHistory tx
```
It's a bit weird that we provide the full history on the `Rollback` event. In a way, this is shifting the full history from the recursive type to shifting the whole history (as a list) here.

I think we should keep this a `ChainStateType` here.

In fact, the single usage site of `Rollback` event in our production code will (via the `ChainRolledBack` event) just use `setChainStateHistory` and that function only uses `currentState` of it!

ch1bo replied:
This will further complicate the `recoverChainStateHistory` function, but if you think about it, the information is already present in the persisted `StateChanged` events.

We originally only had a `Rollback ChainSlot` event here and maybe this is the most minimal we can get by again?

There is a trade-off with my suggestion: we would trade redundant information in the event store with duplicating the logic for "how to rollback". That is, walking into the past and finding the first `ChainStateType` with `<= rollbackSlot`

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/State.hs:
```diff
@@ -189,6 +188,14 @@ instance HasKnownUTxO ChainState where
+  ChainStateAt
+    { chainState = Idle
+    , recordedAt = Nothing
+    }
```
As we are refactoring the chain state handling here, I think we should get rid of in the same go.

The `Idle` chain state does not contain any information. Neither does such a chain state exist for other chain instances (`SimpleTx`) and we can get rid of this completely by making the `ChainStateHistory` a normal list and starting it with `[]`.

ffakenz replied:
after an exploration we discovered that:

- Idle actually records information that we are in Idle just after Abort and Fanout.

- This is in fact the reason why we can't just say we should never be on a Maybe ChainState.

ffakenz replied:
Leaving the refactor of making the `ChainStateHistory` a normal list and starting it with `[]` for another PR.

ch1bo reviewed (changes requested):
I like the direction this is going. The fact that we do not need the full history on every chain event is great.

I do have two major points though, which are worth looking into.

As we are refactoring here, I feel like we should be pushing further and directly to a better situation. Details below :point_down:

ffakenz committed:
Update CHANGELOG

ffakenz committed:
Alter Rollback chain event to not carry the rolledback history

Instead, we made a generic rollbackHistory function, that uses the chainStateSlot
to know until which point in time to rollback from history.

This brings the benefit of making ChainRolledBack very lightweight, reducing the
persistence size.

ffakenz committed:
Remove unnecessary function over ChainStateHistory

This avoids the need to expose the underlying structure of it.

ffakenz committed:
Fix typo

ffakenz committed:
Remove unused function getChainState

ch1bo commented on CHANGELOG.md:
```diff
@@ -17,6 +17,10 @@ changes.
 - Remove hydra-tools package. Move functionality to generate hydra keys to the
   hydra-node executable.
 
+- **BREAKING** Changes to `hydra-node` state persistency:
```
The fact that the golden tests are only removing `previous` and still pass, is proof that this is in fact a backward compatible change! Well done!

ch1bo commented on hydra-node/src/Hydra/Chain.hs:
```diff
@@ -171,6 +173,43 @@ instance Arbitrary Lovelace where
 instance (IsTx tx, Arbitrary (ChainStateType tx)) => Arbitrary (PostTxError tx) where
   arbitrary = genericArbitrary
 
+newtype ChainStateHistory tx = UnsafeChainStateHistory (NonEmpty (ChainStateType tx))
```
Minor: Some documentation what this exactly is and what it should be used for would be great.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -67,40 +73,42 @@ import System.IO.Error (userError)
  where
+  getLatest tv = do
+    chainStateHistory <- readTVar tv
+    return (currentState chainStateHistory)
```
Minor: We use `pure` over `return` throughout the codebase. Not sure if you have a strong opinion on this, I don't. We could avoid altogether her by using `fmap`.

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -67,40 +73,42 @@ import System.IO.Error (userError)
   }
 
 -- | Initialize a new local chain state with given 'ChainStateAt' (see also
 -- 'initialChainState').
```
Should: Update documentation when it gets outdated through signature changes

ch1bo commented on hydra-node/src/Hydra/HeadLogic.hs:
```diff
@@ -725,7 +728,7 @@ aggregate st = \case
         , pendingCommits = Set.fromList parties
         , committed = mempty
-        , chainState
+        , chainState = chainState
```
Minor: no need to put an assignment here and below. The punning (as it was before) would update the field already.

ch1bo commented on hydra-node/src/Hydra/HeadLogic/State.hs:
```diff
@@ -13,6 +13,8 @@ import Hydra.Party (Party, deriveParty)
 import Hydra.Snapshot (ConfirmedSnapshot, Snapshot (..), SnapshotNumber)
 
+import Test.QuickCheck.Instances.Semigroup ()
+
```
I think this is not needed here.

ffakenz replied:
good catch!

ch1bo commented on hydra-node/test/Hydra/BehaviorSpec.hs:
```diff
@@ -613,13 +615,15 @@ simulatedChainAndNetwork initialChainState = do
+      cs <- currentState <$> readTVar chainStateVar
+      let cs' = advanceSlot cs
+      modifyTVar' chainStateVar $ \(UnsafeChainStateHistory prev) ->
+        UnsafeChainStateHistory (cs' :| tail prev)
```
As we have a full chain state history here, I think we should just `pushNewState` here.

ch1bo replied:
Thinking about this further, I wonder why we don't just use the `LocalChainState m` handle here as well? (That just packs a history into a tvar).

EDIT: We can't do that right away because `LocalChainState` is fixed for `Tx` and this code works on a generic `tx`. However, if we change the interface of `LocalChainState.rollback` to be using the generic `ChainSlot` as argument, we could stay abstract and use it here as well.

ffakenz replied:
Changes made:
- rollback is now using the generic `ChainSlot` as argument
- LocalChainState is now abstract over tx

ch1bo commented on hydra-node/src/Hydra/Chain/Direct/Handlers.hs:
```diff
@@ -67,40 +73,42 @@ import System.IO.Error (userError)
 data LocalChainState m = LocalChainState
   { getLatest :: STM m ChainStateAt
   , pushNew :: ChainStateAt -> STM m ()
```
To make this consistent, we should also use `ChainStateType Tx` in the `getLatest` and `pushNew` functions.

ch1bo reviewed (approved):
This is overall a very nice refactoring! Only minor points about outdated docs etc. and please update the changelog to indicate this is not a breaking change. Good job!

ffakenz committed:
Remove BREAKING change tag from CHANGELOG for this refactor

The fact that the golden tests are only removing previous and still pass,
is proof that this is in fact a backward compatible change.

ffakenz committed:
Improve LocalChainState interface

Make it depend on the type family instead of the concrete instance.

Also remove unused imports and minor formatting changes.

ffakenz committed:
Remove dependency on Cardano.Api for LocalChainState

ffakenz committed:
Keep the default chain state as part of the history.

This removes the need of passing the initial chain state around and
allows us to use the LocalChainState handle on the BehaviorSpec.

ffakenz committed:
Add haddocks for chain state history and update outdated ones

ffakenz committed:
Fix logs definitions for StateChanged
